runtime/src/accounts_index.rs

```diff
@@ -1,38 +1,297 @@
+    #[borrows(lock)]
+    pub guard: RwLockWriteGuard<'this, T>,
+}
+
+impl<T> Deref for OwningWriteGuard<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &*self.guard
+    }
+}
+
+impl<T> DerefMut for OwningWriteGuard<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.guard
+    }
+}
 
 pub type SlotList<T> = Vec<(Slot, T)>;
 pub type SlotSlice<'s, T> = &'s [(Slot, T)];
 pub type Ancestors = HashMap<Slot, usize>;
 
 pub type RefCount = u64;
-type AccountMapEntry<T> = (AtomicU64, RwLock<SlotList<T>>);
 pub type AccountMap<K, V> = BTreeMap<K, V>;
 
+type AccountMapEntry<T> = Arc<AccountMapEntryInner<T>>;
+
+#[derive(Debug)]
+pub struct AccountMapEntryInner<T> {
+    ref_count: AtomicU64,
+    pub slot_list: RwLock<SlotList<T>>,
+}
+
+#[self_referencing]
+pub struct ReadAccountMapEntry<T: 'static> {
+    pub owned_entry: AccountMapEntry<T>,
+    #[borrows(owned_entry)]
+    pub slot_list_guard: RwLockReadGuard<'this, SlotList<T>>,
+}
+
+impl<T: Clone> ReadAccountMapEntry<T> {
+    pub fn from_account_map_entry(account_map_entry: AccountMapEntry<T>) -> Self {
+        ReadAccountMapEntryBuilder {
+            owned_entry: account_map_entry,
+            slot_list_guard_builder: |lock| lock.slot_list.read().unwrap(),
+        }
+        .build()
+    }
+
+    pub fn slot_list(&self) -> &SlotList<T> {
+        &*self.slot_list_guard
+    }
+
+    pub fn ref_count(&self) -> &AtomicU64 {
+        &self.owned_entry.ref_count
+    }
+}
+
+#[self_referencing]
+pub struct WriteAccountMapEntry<T: 'static> {
+    pub owned_entry: AccountMapEntry<T>,
+    #[borrows(owned_entry)]
+    pub slot_list_guard: RwLockWriteGuard<'this, SlotList<T>>,
+}
+
+impl<T: 'static + Clone> WriteAccountMapEntry<T> {
+    pub fn from_account_map_entry(account_map_entry: AccountMapEntry<T>) -> Self {
+        WriteAccountMapEntryBuilder {
+            owned_entry: account_map_entry,
+            slot_list_guard_builder: |lock| lock.slot_list.write().unwrap(),
+        }
+        .build()
+    }
+
+    pub fn slot_list_mut(&mut self) -> &mut SlotList<T> {
+        &mut *self.slot_list_guard
+    }
+
+    pub fn ref_count(&self) -> &AtomicU64 {
+        &self.owned_entry.ref_count
+    }
+
+    // Try to update an item in account_maps. If the account is not
+    // already present, then the function will return back Some(account_info) which
+    // the caller can then take the write lock and do an 'insert' with the item.
+    // It returns None if the item is already present and thus successfully updated.
+    pub fn update(&mut self, slot: Slot, account_info: T, reclaims: &mut SlotList<T>) {
+        // filter out other dirty entries from the same slot
+        let mut same_slot_previous_updates: Vec<(usize, (Slot, &T))> = self
+            .slot_list_mut()
+            .iter()
+            .enumerate()
+            .filter_map(|(i, (s, value))| {
+                if *s == slot {
+                    Some((i, (*s, value)))
```

Comment:
nits: this looks so similar to the enclosing closure signature. Can we use `filter` instead of `filter_map` somehow? Because this is fairly in the tight loop. I want to squeeze out anything unneeded as much as possible. :)

Reply:
yeah for sure, removed!